### DISCUSSION TITLE
New multi-settings

### DISCUSSION THREAD

**Discussion by moalamri** [moalamri/vscode-inline-fold#55] [Polls]
## Expanded
```jsonc
{
"[html]":
 {
   "inlineFold.settings" : [
       {
         "pre": "class=",
         "start": "\"",
         "end" : "\""
       },
       {
         "pre": "attribute=",
         "start": "{",
         "end" : "}"
       },
    ] 
 }
}
```

## Condensed
```jsonc
{
"[html]":
 {
   "inlineFold.settings" : 
       { 
         // by order
         "pre": ["class=", "attribute="],
         "start": ["\"", "{"],
         "end" : ["\"", "}"]
       }
 }
}
```

## Advanced
```jsonc
{
"[html]":
 {
   "inlineFold.settings" : 
       { 
         "regex": "/^[a-z]",
         "regexFlags" : "g",
         "regexGroup": 2,
       }
 }
}
```

**Comment by emedeirox** [comment]
The (pre, start & end) atributtes is because #24 ?

I am  between Expanded and Andvanced, the 2nd can be hard to read if it grows too much
and the new settings: Editor > Experimental > Sticky Scroll helps to increase readbility.
The use of array in Expanded settings will be helpful for #34  , but i like the idea of making my own regexs.

**Reply by moalamri** [reply]
Good points!

As for the first suggestion, it will be regex-free (faster and accurate) meaning #24 will not be an issue.
As for the `Expanded` you have a good point that it would probably look so ugly when it grows, also could easily produce mistakes by user mis-ordering, because the `start` character must match the `end` in order.
The `Advanced` won't solve #46 though 😞

**Comment by emedeirox** [comment]
I wonder if possible something like this.
A Disclaimer would be nice for while for those who adventure with regex so they'll be aware of #46.

## Mixed
```json
{
    "[default]":{ 
       "inlineFold.settings" : [
            {
             "pre": "class=",
             "start": "\"",
             "end" : "\""
            },
            {
             "pre": "attribute=",
             "start": "{",
             "end" : "}"
            },
            {
             "regex": "<svg(\\s*.*?\\s*)<\/svg>",
             "regexFlags" : "g",
             "regexGroup": 2,
             //optional atributes overrides settings only for this case (if present)
             "maskChar": ">✧✧✧",
             "maskColor":"#CF9602",
             "unfoldedOpacity":0.8
             //optional
            }
        ]
    },
    "[java]":{
        "inlineFold.settings" : [
             {
                "regex": "(public static void main(String[] args))",
                "regexFlags" : "g",
                "regexGroup": 1,
                "maskChar": "!fun main()",// definitively not fun
             }
         ]
     }
}
```

**Reply by moalamri** [reply]
A mixin solution is possible, but what is the preferred order? 
The old way settings will remain so the current user will not have to change anything, but should I move the old settings to be also language scoped as well? Or it might be a fallback to the new settings where the extension will first look for any lang settings if not then it will use the old global settings, What do you think?

**Reply by moalamri** [reply]
Good that you mentioned #46
I made the parser goes this way:
Look for elements ->
determine if the element tag is closed at the same line or not.
If closed then it will start to breakdown attributes and their values (without props evaluation) to the tag closing character.
If not closed then the same process to the newline character.
Last step is matching settings with what being parsed.

That being said it will only work with the new settings "pre", "start" and "end"

**Reply by emedeirox** [reply]
The hierarchic order I think should be implemented is this:
1. New Settings
   - pre, start end(I remembered you said it has better performance)
     - Language scoped
     - Global
   - regex:  (if not conflicts with the previous)
     - Language scoped
     - Global
2. Old Settings (Not Sure but if it could be Deprecated, maybe in case change configuration page in future)

If the deprecation of setting occur  the current setting should be moved to regex-> global (last).

**Reply by moalamri** [reply]
@e-medeiros Well said 👍
It is better in that order. Thank you for your suggestion.

I've reported a bug to VS Code team that involves an API used in this extension, if they fixed it by this month's release then I'll remove the debouncing class that I implemented which will give much of room to implement more logics of settings heritage and the jsx parser.
The de-bouncer class alone is 500ms delay :/

**Reply by moalamri** [reply]
@e-medeiros Please check [jsx.ts](https://github.com/moalamri/vscode-inline-fold/blob/dev/0.2.0/src/helpers/jsx.ts)
I'm not sure if I'm using the correct names, could you please check if you have used react before?

**Reply by emedeirox** [reply]
@moalamri Sorry but I don't know how to help. Never used react. :(

**Reply by moalamri** [reply]
@e-medeiros And please never use it 😂
I'll get the correct name from either prisma or any other compiler. Thanks 👍

**Reply by moalamri** [reply]
@e-medeiros looks like the poll put the expanded options as the best followed by condensed. I can't agree less as readability is important than the space.
